File: src/main/java/org/endeavourhealth/imapi/model/qof/QOFExpressionNode.java
package org.endeavourhealth.imapi.model.qof;

import java.util.ArrayList;
import java.util.List;

public class QOFExpressionNode {
  private String operator;
  private String condition;
  private String passResult;
  private String failResult;
  private List<QOFExpressionNode> children;

  private QOFExpressionNode() {
    this.children = new ArrayList<>();
  }

  // Factory methods to avoid constructor ambiguity
  public static QOFExpressionNode createOperatorNode(String operator) {
    QOFExpressionNode node = new QOFExpressionNode();
    node.operator = operator;
    return node;
  }

  public static QOFExpressionNode createConditionNode(String condition) {
    QOFExpressionNode node = new QOFExpressionNode();
    node.condition = condition;
    return node;
  }

  public String getOperator() {
    return operator;
  }

  public void setOperator(String operator) {
    this.operator = operator;
  }

  public String getCondition() {
    return condition;
  }

  public void setCondition(String condition) {
    this.condition = condition;
  }

  public List<QOFExpressionNode> getChildren() {
    return children;
  }

  public void addChild(QOFExpressionNode child) {
    this.children.add(child);
  }

  public String getPassResult() {
    return passResult;
  }

  public QOFExpressionNode setPassResult(String result) {
    this.passResult = result;
    return this;
  }

  public String getFailResult() {
    return failResult;
  }

  public QOFExpressionNode setFailResult(String failResult) {
    this.failResult = failResult;
    return this;
  }

  public String toFormattedString() {
    return toRecursiveFormattedString(0)
      .trim()+"\n"
      + "On Pass: " + this.passResult+"\n"
      + "On Fail: " + this.failResult;
  }
  private String toRecursiveFormattedString(int indent) {
    StringBuilder sb = new StringBuilder();
    String indentStr = "    ".repeat(indent);

    if (operator != null && !operator.isEmpty()) {
      sb.append(indentStr).append(operator).append("\n");
      for (QOFExpressionNode child : children) {
        sb.append(child.toRecursiveFormattedString(indent + 1)).append("\n");
      }
    } else if (condition != null && !condition.isEmpty()) {
      sb.append(indentStr).append(condition);
    }

    return sb.toString();
  }
}

File: src/main/java/org/endeavourhealth/imapi/model/qof/QOFExpressionNode.kt
package org.endeavourhealth.imapi.model.qof

const val ANSI_RESET: String = "\u001B[0m"
const val ANSI_RED: String = "\u001B[31m"
const val ANSI_GREEN: String = "\u001B[32m"
const val ANSI_YELLOW: String = "\u001B[33m"
const val ANSI_BLUE: String = "\u001B[34m"
const val ANSI_PURPLE: String = "\u001B[35m"
const val ANSI_CYAN: String = "\u001B[36m"

class QOFExpressionNode {
  val condition: QOFCondition = QOFCondition()
  var operator: String? = null
  var passResult: String? = null
  var failResult: String? = null
  val children: MutableList<QOFExpressionNode> = mutableListOf()

  fun toFormattedString(): String {
    val result = toRecursiveFormattedString(0).trim()

    val finalResult = StringBuilder(result)
    if (passResult != null) {
      finalResult.append("\n${ANSI_GREEN}Pass: $passResult $ANSI_RESET")
    }
    if (failResult != null) {
      finalResult.append("\n${ANSI_RED}Fail: $failResult $ANSI_RESET")
    }

    return finalResult.toString()
  }

  private fun toRecursiveFormattedString(indent: Int): String {
    val sb = StringBuilder()
    val indentStr = "    ".repeat(indent)

    if (!operator.isNullOrEmpty()) {
      sb.append("$indentStr$ANSI_YELLOW$operator$ANSI_RESET\n")
      for (child in children) {
        val childResult = child.toRecursiveFormattedString(indent + 1)
        sb.append(childResult)
      }
    } else if (!condition.leftOperand.isNullOrEmpty()) {
      sb.append("$indentStr$ANSI_BLUE${condition.leftOperand}$ANSI_RESET $ANSI_PURPLE${condition.comparator}$ANSI_RESET $ANSI_CYAN${condition.rightOperand}\n")
    }

    val result = sb.toString()

    return if (result.replace("\n", "").isBlank()) "==EMPTY==" else result
  }
}
